### PR TITLE
Add map of class to list of dataset index

### PR DESCRIPTION
**Pull request by matanper**
#### Reference Issues/PRs

solves num. 2 in #1022 
and also num. 1 of Heatmap Comparison and Simple Model Comparison

#### What does this implement/fix? Explain your changes.

Instead of caching the number of samples per class, caching for each class the batch indices of samples that have this class (in the label).
Also added the the cache update to the larger update loops, so only remains to remove the ad-hoc calculation after removing the usage of it from the init_run of checks

### CODE REVIEW

File: deepchecks/vision/vision_data.py [hunk 1]
```diff
@@ -255,9 +278,11 @@ def copy(self, n_samples: int = None, shuffle: bool = False, random_state: int =
                                                                   shuffle=shuffle,
                                                                   random_state=random_state,
                                                                   n_samples=n_samples)
-        # If new data is sampled, then needs to remove cached info
+        # If new data is sampled, then needs to re-calculate cache
         if n_samples:
-            new_vision_data._n_of_samples_per_class = None
+            new_vision_data.init_cache()
+            for batch in new_vision_data:
+                new_vision_data.update_cache(self.batch_to_labels(batch))
```
Comment (noamzbr) [inline deepchecks/vision/vision_data.py:285]: When does this happen? because we use copy to create the sampled data obj?

Comment (matanper) [reply deepchecks/vision/vision_data.py:285]: yes

File: deepchecks/vision/classification_data.py [hunk 2]
```diff
@@ -123,7 +123,7 @@ def infer_on_batch(self, batch, model, device) -> Union[List[torch.Tensor], torc
 
     def get_classes(self, batch_labels: Union[List[torch.Tensor], torch.Tensor]):
         """Get a labels batch and return classes inside it."""
-        return batch_labels.tolist()
+        return batch_labels.reshape(-1, 1).tolist()
```
[the remark this answers is not in the document]
Comment (matanper) [reply deepchecks/vision/classification_data.py:126]: Looks like the tests are passing, it is in order to return list per sample

File: deepchecks/vision/checks/performance/simple_model_comparison.py [hunk 3]
```diff
@@ -109,39 +109,13 @@ def __init__(self,
...
-                self._state['dummy_prediction_generator'] = \
-                    lambda: torch.from_numpy(np.random.multinomial(1, class_prior))
-            elif self.strategy == 'uniform':
-                self._state['dummy_prediction_generator'] = \
-                    lambda: torch.from_numpy(np.ones(context.train.num_classes) / context.train.num_classes)
-            else:
-                raise DeepchecksValueError(
-                    f'Unknown strategy type: {self.strategy}, expected one of {_allowed_strategies}.'
-                )
+        self._test_metrics = get_scorers_list(context.train, self.alternative_metrics)
```
Comment (ItayGabbay) [inline deepchecks/vision/checks/performance/simple_model_comparison.py:118]: Can you just add a simple if to check whether the strategy is valid?

Comment (matanper) [reply deepchecks/vision/checks/performance/simple_model_comparison.py:118]: This `if` exists in the init

**Review by ItayGabbay — APPROVED** [review]

**Pull request merged**